chunkset_tpl.h

```diff
@@ -112,66 +67,77 @@ Z_INTERNAL uint8_t* CHUNKMEMSET(uint8_t *out, unsigned dist, unsigned len) {
        Assert(len >= sizeof(uint64_t), "chunkmemset should be called on larger chunks"); */
     Assert(dist > 0, "chunkmemset cannot have a distance 0");
 
-    unsigned char *from = out - dist;
-    chunk_t chunk;
-    unsigned sz = sizeof(chunk);
-    if (len < sz) {
-        while (len != 0) {
-            *out++ = *from++;
-            --len;
-        }
-        return out;
-    }
+    uint8_t *from = out-dist;
 
-#ifdef HAVE_CHUNKMEMSET_1
     if (dist == 1) {
-        chunkmemset_1(from, &chunk);
-    } else
-#endif
+        memset(out, *from, len);
+        return out + len;
+    } else if (dist > sizeof(chunk_t)) {
+        out = CHUNKUNROLL(out, &dist, &len);
+        return CHUNKCOPY(out, out - dist, len);
+    }
+
+    chunk_t chunk_load;
+
+    /* TODO: possibly build up a permutation table for this if not an even modulus */
 #ifdef HAVE_CHUNKMEMSET_2
     if (dist == 2) {
-        chunkmemset_2(from, &chunk);
+        chunkmemset_2(from, &chunk_load);
     } else
 #endif
 #ifdef HAVE_CHUNKMEMSET_4
     if (dist == 4) {
-        chunkmemset_4(from, &chunk);
+        chunkmemset_4(from, &chunk_load);
     } else
 #endif
 #ifdef HAVE_CHUNKMEMSET_8
     if (dist == 8) {
-        chunkmemset_8(from, &chunk);
+        chunkmemset_8(from, &chunk_load);
+    } else if (dist == sizeof(chunk_t)) {
+        loadchunk(from, &chunk_load);
     } else
 #endif
-    if (dist == sz) {
-        loadchunk(from, &chunk);
-    } else if (dist < sz) {
-        unsigned char *end = out + len - 1;
-        while (len > dist) {
-            out = CHUNKCOPY_SAFE(out, from, dist, end);
-            len -= dist;
+    {
+        /* This code takes string of length dist from "from" and repeats
+         * it for as many times as can fit in a chunk_t (vector register) */
+        int32_t cpy_dist;
+        int32_t bytes_remaining = sizeof(chunk_t);
+        uint8_t *cur_chunk = (uint8_t*)&chunk_load;
+        while (bytes_remaining) {
+            cpy_dist = MIN(dist, bytes_remaining);
+            memcpy(cur_chunk, from, cpy_dist);
+            bytes_remaining -= cpy_dist;
+            cur_chunk += cpy_dist;
         }
-        if (len > 0) {
-            out = CHUNKCOPY_SAFE(out, from, len, end);
+    }
+
+    /* If we're lucky enough and dist happens to be an even modulus of our vector length,
+     * we can do two stores per loop iteration, which for most ISAs, especially x86, is beneficial */
+    uint32_t chunk_mod = sizeof(chunk_t) % dist;
+    if (chunk_mod == 0) {
```

Comment:
Question, why do we care if `chunk_mod == 0` when performing the while loop below. Say if `dist = 91`, it would return `chunk_mod > 0` but that loop could still be run.. just up until `len < 2 * sizeof(chunk_t)`. So doesn't it make sense to remove the if statement?

Reply:
Because our dist here needs to be an even sized modulus of our chunk size to allow us to store two adjacent chunks in memory without an adv_amount increment to the pointer in between.  If the destination for the store to memory overlaps, we can't actually get any ILP on the store, anyway, as they both end up conflicting at the destination address.  It's possible store forwarding _might_ be able to save us here, but if I had to wager doing this where things get stored on top of each other is non-ideal.  You could maybe blend things together with blends or shuffles but well, that opens a _new_ can of worms entirely.  For one, it doesn't work with our chunk_t interface we have exposed but for two, you end up basically constructing a blend/shuffle vector that moves after every store, which is definitely not cheaper.

Reply:
Oh and there's also the convenience that sizeof(chunk_t) is known at compile time.  This means the compile can encode the offset from the address more simply and predictably.  The microarchitecture can know the offset of the next store never changes if it's encoded as a hard offset in an instruction and it can easily arm cache lines for the next store.

Reply:
> Because our dist here needs to be an even sized modulus of our chunk size to allow us to store two adjacent chunks in memory without an adv_amount increment to the pointer in between.

Thanks, I think I see what you are saying.